cpp/src/core/core_number.cu

```diff
@@ -0,0 +1,141 @@
+  gdf_error err = gdf_add_adj_list(graph);
+  if (err != GDF_SUCCESS)
+    return err;
+  GDF_REQUIRE(graph->adjList->offsets->dtype == GDF_INT32, GDF_UNSUPPORTED_DTYPE);
+  GDF_REQUIRE(graph->adjList->indices->dtype == GDF_INT32, GDF_UNSUPPORTED_DTYPE);
+  GDF_REQUIRE(core_number->dtype == GDF_INT32, GDF_UNSUPPORTED_DTYPE);
+  GDF_REQUIRE(core_number->size == graph->numberOfVertices, GDF_COLUMN_SIZE_MISMATCH);
+
+  return core_number_impl(graph, static_cast<int*>(core_number->data));
+}
+
+struct LessThan {
+  int k;
+
+  LessThan(int _k) : k(_k) {}
+
+  __host__ __device__
+    bool operator()(const int core) {
+      return core < k;
+    }
+};
+
+gdf_error gdf_k_core(gdf_graph *in_graph,
+                     int k,
+                     gdf_column *vertex_id,
+                     gdf_column *core_number,
+                     gdf_graph *out_graph) {
+  GDF_REQUIRE(out_graph != nullptr && in_graph != nullptr, GDF_INVALID_API_CALL);
+  GDF_REQUIRE(in_graph->adjList != nullptr || in_graph->edgeList != nullptr, GDF_INVALID_API_CALL);
+  gdf_error err = gdf_add_adj_list(in_graph);
+  if (err != GDF_SUCCESS)
+    return err;
+  GDF_REQUIRE(in_graph->adjList->offsets->dtype == GDF_INT32, GDF_UNSUPPORTED_DTYPE);
+  GDF_REQUIRE(in_graph->adjList->indices->dtype == GDF_INT32, GDF_UNSUPPORTED_DTYPE);
+  GDF_REQUIRE((vertex_id == nullptr) == (core_number == nullptr), GDF_INVALID_API_CALL);
+  if (vertex_id != nullptr) {
+    GDF_REQUIRE(vertex_id->dtype == GDF_INT32, GDF_UNSUPPORTED_DTYPE);
+    GDF_REQUIRE(core_number->dtype == GDF_INT32, GDF_UNSUPPORTED_DTYPE);
+    GDF_REQUIRE(core_number->size == vertex_id->size, GDF_INVALID_API_CALL);
+  }
+
+  cudaStream_t stream {nullptr};
+  gdf_size_type nV = in_graph->numberOfVertices;
+
+  rmm::device_vector<int> vertexIdentifier;
+  rmm::device_vector<int> coreNumber;
+
+  int * vertex_identifier_ptr = nullptr;
+  int * core_number_ptr = nullptr;
+  gdf_size_type vLen = 0;
+
+  if (vertex_id == nullptr) {
+    vertexIdentifier.resize(nV);
+    thrust::sequence(
+        rmm::exec_policy(stream)->on(stream),
+        vertexIdentifier.begin(), vertexIdentifier.end());
+    vertex_identifier_ptr = vertexIdentifier.data().get();
+    vLen = nV;
+
+    coreNumber.resize(nV);
+    core_number_ptr = coreNumber.data().get();
+    err = core_number_impl(in_graph, core_number_ptr);
+    GDF_REQUIRE(err, GDF_SUCCESS);
+  } else {
+    core_number_ptr = static_cast<int*>(core_number->data);
+    vertex_identifier_ptr = static_cast<int*>(vertex_id->data);
+    vLen = vertex_id->size;
+  }
+
+  auto cnPtr = thrust::device_pointer_cast(core_number_ptr);
+  if (k == -1) {
+    k = thrust::reduce(
+        rmm::exec_policy(stream)->on(stream),
+        cnPtr, cnPtr + vLen, -1, thrust::maximum<int>());
+  }
+
+  rmm::device_vector<int> filteredVertices(vLen);
+  auto vInPtr = thrust::device_pointer_cast(vertex_identifier_ptr);
+  auto ptr = thrust::remove_copy_if(
```

Comment:
Ditto above.  Aren't vInPtr and ptr also just int *?

I'm a little less pressed about this since you only use the variable in the very next statement.  But I think int * here (or whatever the type really is if not that) is a little more clear for someone coming along later to understand or modify the code.

Reply:
I can replace `auto` with `thrust::device_ptr<int>` here for `vInPtr`.

`ptr` however is `thrust::detail::normal_iterator<thrust::device_ptr<int>>`. Its a bit too long and unclear so that's why I chose `auto`.